packages/contracts-bedrock/test/properties/medusa/ProtocolAtomic.t.sol

```diff
@@ -0,0 +1,199 @@
+        uint8 remoteTokenIndex;
+        uint8 nameIndex;
+        uint8 symbolIndex;
+        uint8 decimalsIndex;
+    }
+
+    address[] internal remoteTokens;
+    address[] internal allSuperTokens;
+
+    //@notice  'real' deploy salt => total supply sum across chains
+    EnumerableMap.Bytes32ToUintMap internal ghost_totalSupplyAcrossChains;
+
+    constructor() {
+        vm.etch(address(MESSENGER), address(new MockL2ToL2CrossDomainMessenger()).code);
+        superchainERC20Impl = new OptimismSuperchainERC20();
+        for (uint256 i = 0; i < INITIAL_TOKENS; i++) {
+            _deployRemoteToken();
+            for (uint256 j = 0; j < INITIAL_SUPERTOKENS; j++) {
+                _deploySupertoken(remoteTokens[i], WORDS[0], WORDS[0], DECIMALS[0], j);
+            }
+        }
+    }
+
+    /// @notice the deploy params are _indexes_ to pick from a pre-defined array of options and limit
+    /// the amount of supertokens for a given remoteAsset that are incompatible between them, as
+    /// two supertokens have to share decimals, name, symbol and remoteAsset to be considered
+    /// the same asset, and therefore bridgable.
+    modifier validateTokenDeployParams(TokenDeployParams memory params) {
+        params.remoteTokenIndex = uint8(bound(params.remoteTokenIndex, 0, remoteTokens.length - 1));
+        params.nameIndex = uint8(bound(params.nameIndex, 0, WORDS.length - 1));
+        params.symbolIndex = uint8(bound(params.symbolIndex, 0, WORDS.length - 1));
+        params.decimalsIndex = uint8(bound(params.decimalsIndex, 0, DECIMALS.length - 1));
+        _;
+    }
+
+    /// @notice deploy a new supertoken with deploy salt determined by params, to the given (of course mocked) chainId
+    function fuzz_DeployNewSupertoken(
+        TokenDeployParams memory params,
+        uint256 chainId
+    )
+        external
+        validateTokenDeployParams(params)
+    {
+        chainId = bound(chainId, 0, MAX_CHAINS - 1);
+        _deploySupertoken(
+            remoteTokens[params.remoteTokenIndex],
+            WORDS[params.nameIndex],
+            WORDS[params.symbolIndex],
+            DECIMALS[params.decimalsIndex],
+            chainId
+        );
+    }
+
+    /// @custom:property-id 22
+    /// @custom:property sendERC20 decreases sender balance in source chain and increases receiver balance in
+    /// destination chain exactly by the input amount
+    /// @custom:property-id 23
+    /// @custom:property sendERC20 decreases total supply in source chain and increases it in destination chain exactly
+    /// by the input amount
+    function fuzz_SelfBridgeSupertoken(uint256 fromIndex, uint256 destinationChainId, uint256 amount) external {
+        destinationChainId = bound(destinationChainId, 0, MAX_CHAINS - 1);
+        fromIndex = bound(fromIndex, 0, allSuperTokens.length - 1);
+        OptimismSuperchainERC20 sourceToken = OptimismSuperchainERC20(allSuperTokens[fromIndex]);
+        OptimismSuperchainERC20 destinationToken =
```

Comment:
Don't get why the destination token address depends also on the `msg.sender`

Reply:
And also, I think in this way you're limiting the test to check balances the same address on origin and destination (`msg.sender`).
IMO it should be more exhaustive if we check on different addresses as well, but it's ok if that adds even more complexity and isn't worth it

Reply:
In this PR I tried to get medusa running for the simplest possible (but still somewhat meaningful) test. I have #22 where I add cross-chain transfers between different accounts, which I deemed out of scope for this one since it requires an actor setup and this PR is big enough already :grin: